tests/rptest/tests/storage_timing_stress_test.py

```diff
@@ -0,0 +1,252 @@
+# Copyright 2023 Redpanda Data, Inc.
+#
+# Licensed as a Redpanda Enterprise file under the Redpanda Community
+# License (the "License"); you may not use this file except in compliance with
+# the License. You may obtain a copy of the License at
+#
+# https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+
+from rptest.clients.rpk import RpkTool
+from rptest.services.admin import Admin
+from rptest.services.cluster import cluster
+from rptest.services.kgo_verifier_services import KgoVerifierProducer, KgoVerifierSeqConsumer
+from rptest.tests.redpanda_test import RedpandaTest
+from rptest.clients.types import TopicSpec
+from rptest.tests.partition_movement import PartitionMovementMixin
+from ducktape.mark import parametrize
+from rptest.utils.mode_checks import skip_debug_mode
+
+import concurrent.futures
+import random
+import time
+from concurrent.futures import ThreadPoolExecutor
+from datetime import datetime
+
+
+class StorageCheck:
+    def __init__(self, name, check):
+        self._name = name
+        self._check = check
+
+    @property
+    def check(self):
+        return self._check
+
+    @property
+    def name(self):
+        return self._name
+
+
+class StorageTimingStressTest(RedpandaTest, PartitionMovementMixin):
+    """
+    The tests in this class are intended to be generic ~~cloud~~ storage test.
+    It's a copy of CloudStorageTimingStressTest, without cloud storage enabled, done in a effort to trigger trigger bugs like the one described in https://github.com/redpanda-data/redpanda/pull/12176
+    They use a workload that enables all operations on the cloud storage log
+    (appends, truncations caused by retention, compacted segment reuploads and
+    adjacent segment merging. A configurable series of checks are performed
+    at every 'check_interval'. If any of the checks result in an exception, or
+    fail to complete the test will fail.
+
+    The tests can be extended by creating a new check function and registering
+    it in the 'prologue' method.
+    """
+
+    mib = 1024 * 1024
+    message_size = 32 * 1024  # 32KiB
+    log_segment_size = 128 * 1024  # 4MiB
+    produce_byte_rate_per_ntp = 8 * mib  # 8 MiB
+    target_runtime = 5 * 60  # seconds
+    check_interval = 10  # seconds
+    check_timeout = 10  # seconds
+    allow_runtime_overshoot_by = 2
```

Comment:
Here's a few ideas for making this a bit more stressful for the system:
1. reduce segment size to something like 1MiB; you'll also have set `log_segment_size_min` otherwise it will be clamped
2. increase the produce rate; since we are not uploading to cloud storage we can go faster.
3. set `retention_local_trim_interval` to something low too to exercise the new disk management code from Noah
4. set `disk_reservation_percent` and `retention_local_target_percent` such that the test constantly overshoots these limits (again, in order to exercise the new disk management code)

Reply:
need to tune for 3 and 4, but should be there